Add tests for analysis protocol fields (#61)

danielsotirhos opened the pull request.

danielsotirhos commented on src/humancellatlas/data/metadata/api.py:
```diff
@@ -459,7 +459,27 @@ def __init__(self, json: JSON):
+    provenance_update_date: Optional[str]
+    protocol_core_protocol_id: str      # Protocol self.protocol_id ?
+    protocol_type_text: str
+
```
Two of the AnalysisProtocol fields are already fetched by parent classes Protocol and Entity. Should these fields be removed from the AnalysisProtocol class to avoid duplication, or kept in AnalysisProtocol to make sure these fields exist for this class?

hannes-ucsc commented on test/test.py:
```diff
@@ -485,6 +475,30 @@ def to_json(fqid):
 
         self.assertEqual({}, errors)
 
+    def test_sequencing_process_paired_end(self):
```
Can you make the code move a separate commit?

hannes-ucsc replied:
Could you elaborate the change from `'prod'` to `''`?

danielsotirhos replied:
Yes, i'll make the code move a separate commit. I moved the function to group tests that don't use _test_bundle() below _test_bundle().

danielsotirhos replied:
Regarding changing 'prod' to '': deployment = "prod" works for _load_bundle() but not if it needs to download the bundle because dss_client() needs it to be empty for prod.

Longer version: The "deployment" var is passed to _load_bundle().
In _load_bundle() the value of the var (or string "prod" if empty) is copied to a var "canning_directory".
If _load_bundle() needs ot download the bundle, dss_client() is called passing the "deployment" var.
dss_client() uses the "deployment" param to build the swagger url, requiring "deployment" to be empty if production is desired (otherwise you'd get https://dss.prod.data.humancellatlas.org...)

hannes-ucsc replied:
I shouldn't have to modify the call to _load_bundle(). The same arguments should both work for initially downloading a bundle and canning it as well as loading an already canned bundle. Is there a bug lurking that you change is merely working around?

hannes-ucsc commented:
Be sure to create a clean history. The deployment change should be one commit. The method move  should be another (even though I doubt it's necessary). The added test should be a third commit.

danielsotirhos commented:
Rebased & cleaned history. One commit for the added test case, another commit to use the string 'prod' instead of a None type to indicate production deployments, and deleted the commit that moved the unrelated function "test_sequencing_process_paired_end()".

hannes-ucsc commented on src/humancellatlas/data/metadata/helpers/dss.py:
```diff
@@ -92,14 +92,13 @@ def download_file(item):
 
 
-def dss_client(deployment: Optional[str] = None) -> DSSClient:
+def dss_client(deployment: Optional[str] = 'prod') -> DSSClient:
```
`Optional[T]` means `Union[None,T]`

We shouldn't permit `None` …

hannes-ucsc commented on src/humancellatlas/data/metadata/helpers/dss.py:
```diff
@@ -92,14 +92,13 @@ def download_file(item):
+    :param deployment: The name of a DSS deployment like `dev`, `integration` or `staging`.
     """
-    deployment = deployment + "." if deployment else ""
+    deployment = deployment + "." if deployment and deployment != 'prod' else ""
```
… and we shouldn't accept False-ish values.

```suggestion
    deployment = "" if deployment == 'prod' else deployment + "." 
```

Also like to avoid negation as much as possible.

danielsotirhos replied:
Easier to read this way, agreed.

hannes-ucsc commented on test/test.py:
```diff
@@ -143,18 +144,17 @@ def _canned_bundle(self, directory, uuid, version):
             return None, None
 
-    def _load_bundle(self, uuid, version, replica='aws', deployment=None):
+    def _load_bundle(self, uuid, version, replica='aws', deployment='prod'):
```
```suggestion
   
    prod = 'prod'

    def _load_bundle(self, uuid, version, replica='aws', deployment=prod):
```

danielsotirhos replied:
I am unsure of the benefit of this suggested change.

hannes-ucsc commented:
Commit titles appear to be the same but they should be different.
